wfe2/wfe.go

```diff
@@ -1204,15 +1208,55 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
 		return
 	}
 
-	response.Header().Set("Content-Type", "application/pem-certificate-chain")
-	response.WriteHeader(http.StatusOK)
-
-	pem := pem.EncodeToMemory(&pem.Block{
+	leafPEM := pem.EncodeToMemory(&pem.Block{
 		Type:  "CERTIFICATE",
 		Bytes: cert.DER,
 	})
-	// TODO(#3291): Serve entire chain instead of just end-entity certificate.
-	if _, err = response.Write(pem); err != nil {
+
+	var responsePEM []byte
+
+	// If the WFE is configured with CertificateChains, construct a chain for this
+	// certificate using its AIA Issuer URL.
+	if len(wfe.CertificateChains) > 0 {
+		parsedCert, err := x509.ParseCertificate(cert.DER)
+		if err != nil {
+			// If we can't parse one of our own certs there's a serious problem
+			wfe.sendError(response, logEvent, probs.ServerInternal(
+				fmt.Sprintf(
+					"unable to parse Boulder issued certificate with serial %#v",
+					serial),
+			), err)
+			return
+		}
+
+		// TODO(@cpu): At present we only use the **first** AIA Issuer URL on the
+		// certificate when constructing the chain. Open question: should we use
+		// _all_ of them? If so, in what order?
+		aiaIssuerURL := parsedCert.IssuingCertificateURL[0]
+		if chain, ok := wfe.CertificateChains[aiaIssuerURL]; ok {
+			responsePEM = append([]byte(chain), leafPEM...)
```

Comment:
I'm not sure of the exact Go semantics here, but I think what you want instead is:

```
			copy(responsePEM, []byte(chain))
			responsePEM = append(responsePEM, leafPEM...)
```

And in fact making that change locally causes the HEAD test to pass. There's still a lot of mysteries here about why just that one thing fails.

Reply:
This works because it just reverts the test back to what it used to do. `copy(responsePEM, []byte(chain))` won't copy anything into `responsePEM` because its size is 0.

Reply:
Good point.

Reply:
Conclusion here was: The Go http library will add Content-Length automatically if the body written is below a certain size. In the old unittest, the output was below that size. Adding the chain (and PEM encoding) pushed the response above that size, so Go does not automatically compute the Content-Length. Solution: Add a content-length header ourselves.

Reply:
> Conclusion here was: The Go http library will add Content-Length automatically if the body written is below a certain size. 

That was a surprising conclusion :-) Hat tip to @rolandshoemaker for figuring that one out! Resolved in 6fd4fb3